### PR TITLE
[+]Server Connection with Frontend

### PR DESCRIPTION
**Pull request by MostlyKIGuess**
# Backend

## Contour_filtering.py
- Biggest issue in this world is this file, model works fine, we need to figure out how to add margins to the images and separate them out better.
- No need for threshold here so removed it.

### resnet18_vgg16.py
- Added more models for further training purposes.

### Server.py
- was mostly done by @0x0is1 . And is left as it is.

# Frontend

### canvas.js
- Removed unnecessary overlay canvas. Set line default to 2 for better results.

### App.js
- checkPrediction function added which will loop over the request after the image is uploaded by the user.

### CODE REVIEW

File: Backend/request_handler/server.py [hunk 1]
```diff
@@ -21,30 +61,85 @@ def get_home():
...
+            gray_image = cv2.cvtColor(resized_image, cv2.COLOR_BGR2GRAY)
+        else:
+            gray_image = resized_image
+
+        _, bw_image = cv2.threshold(gray_image, 220, 256, cv2.THRESH_BINARY)
+
+        cv2.imwrite(image_path, bw_image)
+        prediction = classifier.predict(image_path=image_path)
+        prediction_text += prediction
+    return prediction_text
```
Comment (0x0is1) [inline Backend/request_handler/server.py]: Looks good.

**Review by 0x0is1 — APPROVED** [review]

**Comment by 0x0is1** [comment]
@MostlyKIGuess can i expect change in latex db thing and review notes by today?

**Comment by MostlyKIGuess** [comment]
> @MostlyKIGuess can i expect change in latex db thing and review notes by today?

Done, Check the contour filter once!

**Comment by 0x0is1** [comment]
@MostlyKIGuess changes requires testing.

File: Backend/trainmodel.py [hunk 2]
```diff
@@ -20,6 +24,34 @@
...
+            image = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
+            resized_image = cv2.resize(image, (45, 45), interpolation=cv2.INTER_AREA)
+            if len(resized_image.shape) == 3:
+                gray_image = cv2.cvtColor(resized_image, cv2.COLOR_BGR2GRAY)
+            else:
+                gray_image = resized_image
+
+            _, bw_image = cv2.threshold(gray_image, 220, 256, cv2.THRESH_BINARY)
+            count+=1
+            cv2.imwrite(image_path, bw_image)
```
Comment (0x0is1) [inline Backend/trainmodel.py]: Physical dataset writing wasnt required and is considered a bad practice.

File: Backend/trainmodel.py [hunk 3]
```diff
@@ -1,11 +1,15 @@
...
 from PIL import Image
-
 from learner.resnet18_vgg16 import HandwrittenSymbolsClassifier
 
+
+
+
+
+
+
```
Comment (0x0is1) [inline Backend/trainmodel.py:8]: Yeah a bunch of whitespaces helps.

**Review by 0x0is1 — CHANGES_REQUESTED** [review]

File: Backend/learner/resnet18_vgg16.py [hunk 4]
```diff
@@ -43,17 +45,30 @@ def __init__(self, root_dir, transform=None):
...
             return len(self.images)
         
         def __getitem__(self, idx):
-            image = Image.open(self.images[idx]).convert('RGB') 
+            image_path = self.images[idx]
+            image = cv2.imread(image_path)
+            image = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
+            resized_image = cv2.resize(image, (45, 45), interpolation=cv2.INTER_AREA)
+            _, bw_image = cv2.threshold(resized_image, 220, 255, cv2.THRESH_BINARY)
+            image = Image.fromarray(bw_image).convert('RGB')  # Convert to PIL Image and RGB mode
```
Comment (0x0is1) [inline Backend/learner/resnet18_vgg16.py]: Replaced dataset tuning without touching datasets.

File: Backend/trainmodel.py [hunk 5]
```diff
@@ -60,23 +57,29 @@ def is_image_file(filename):
...
 for image_path in image_paths:
+    if not is_image_file(image_path):
+        continue
+
     image = cv2.imread(image_path)
+
+    if image is None:
+        print(f"Failed to load image: {image_path}")
+        continue
+
```
Comment (0x0is1) [inline Backend/trainmodel.py]: This wasnt required as this is already done in trainer module.

**Review by 0x0is1 — CHANGES_REQUESTED** [review]

**Comment by 0x0is1** [comment]
Fuck i reverted wrong commit

**Comment by MostlyKIGuess** [comment]
bro you literally removed the best commit 😭

**Comment by MostlyKIGuess** [comment]
Make the Electron PR today @0x0is1 ?

File: Backend/trainmodel.py [hunk 6]
```diff
@@ -57,29 +28,21 @@ def is_image_file(filename):
...
     resized_image = cv2.resize(image, (45, 45), interpolation=cv2.INTER_AREA)
     if len(resized_image.shape) == 3:
         gray_image = cv2.cvtColor(resized_image, cv2.COLOR_BGR2GRAY)
     else:
         gray_image = resized_image
 
-    _, bw_image = cv2.threshold(gray_image, 220, 255, cv2.THRESH_BINARY)
-    
-    pil_image = Image.fromarray(bw_image).convert('RGB')
-    prediction = classifier.predict(image_path=pil_image)
```
Comment (0x0is1) [inline Backend/trainmodel.py]: Thanks for fixing the fuckups.

**Review by 0x0is1 — APPROVED** [review]

**Comment by 0x0is1** [comment]
> Make the Electron PR today @0x0is1 ?

Yes, Sure.

**Pull request merged**